src/gemdat/path.py

```diff
@@ -211,22 +225,64 @@ def optimal_path(
         Coordinates of the starting point
     end: tuple
         Coordinates of the ending point
+    method : str
+        Method used to calculate the shortest path. Options are:
+        - 'dijkstra': Dijkstra's algorithm
+        - 'bellman-ford': Bellman-Ford algorithm
+        - 'minmax-energy': Minmax energy algorithm
+        - 'dijkstra-exp': Dijkstra's algorithm with exponential weights
 
     Returns
     -------
     path: Pathway
         Optimal path on the graph between start and end
     """
 
-    optimal_path = nx.shortest_path(F_graph,
-                                    source=start,
-                                    target=end,
-                                    weight='weight')
-    path_energy = [F_graph.nodes[node]['energy'] for node in optimal_path]
+    optimal_path = nx.shortest_path(
+        F_graph,
+        source=start,
+        target=end,
+        weight='weight_exp' if method == 'dijkstra-exp' else 'weight',
+        method='dijkstra'
+        if method == 'dijkstra-exp' or method == 'minmax-energy' else method)
+
+    if method == 'dijkstra' or method == 'bellman-ford' or method == 'dijkstra-exp':
+
+        path_energy = [F_graph.nodes[node]['energy'] for node in optimal_path]
+        path = Pathway(sites=optimal_path, energy=path_energy)
+        return path
+
+    elif method == 'minmax-energy':
+        max_energy = max(
+            [F_graph.nodes[node]['energy'] for node in optimal_path])
+        minmax_energy = max_energy
+        pruned_F_graph = F_graph.copy()
+        while minmax_energy <= max_energy:
+            # Find the node of the path with the highest energy
+            max_node = max(optimal_path,
+                           key=lambda x: F_graph.nodes[x]['energy'])
+            # remove this node from the graph
+            pruned_F_graph.remove_node(max_node)
+            # recompute the path
+            pruned_path = nx.shortest_path(
+                pruned_F_graph,
+                source=start,
+                target=end,
+                weight='weight',
+            )
+            minmax_energy = max(
+                [F_graph.nodes[node]['energy'] for node in pruned_path])
+
+            if minmax_energy < max_energy:
+                optimal_path = pruned_path
+                max_energy = minmax_energy
```

Comment:
If possible, consider refactoring this to get the minmax-energy code 'out of the way'. This will also make it easier to unit test this bit.

You could then re-use the code to construct Pathway in one go I believe.

```suggestion
    if method == 'minmax-energy':
        optimal_path = _optimal_path_minmax_energy(...)
    else:
        raise ValueError(f'Unknown method {method}')
        
    path_energy = [F_graph.nodes[node]['energy'] for node in optimal_path]
    path = Pathway(sites=optimal_path, energy=path_energy)
    return path
```

Reply:
Thanks, I have implemented this suggestion. If you agree with the last commit, I think we can merge this branch.